File: src/readers/osm/primitive.ts
import { Way } from './way';
import { DenseNodes, Node } from './node';
import { Relation, getNodeRelationPairs } from './relation';

import type { IntermediateNodeMember } from './relation';
import type { Pbf as Protobuf } from '../../readers/protobuf';
import type { InfoBlock, OSMReader } from '.';

/** The expected metadata in the VectorFeature for all types (node, way, relation) */
export interface Metadata {
  info: InfoBlock;
  nodes?: IntermediateNodeMember[];
  relation?: {
    role: string;
    properties: Record<string, string>;
  };
}

/**
 * NOTE: currently relations are stored, but we don't wait for the Block to store all relations
 * before we start testing primtiveHandle against the data. This is a problem because
 * relations reference eachother at times, and we need to be able to resolve those references
 * before we can run relationHandle against the data. This isn't an important issue since
 * in practice, all relations that reference eachother often produce garbage or unusable data.
 * But it would be *nice* to fix this. Morbidly enough, the "BEST" solution is to treat relations
 * like we do nodes and ways since relations could possibly reference eachother outside their own block.
 * From a practical standpoint, I can't see this being worth the effort or memory/time cost.
 */
export class PrimitiveBlock {
  stringtable!: StringTable;
  primitiveGroups: PrimitiveGroup[] = [];
  // Granularity, units of nanodegrees, used to store coordinates in this block.
  granularity = 100;
  // Offset value between the output coordinates and the granularity grid in units of nanodegrees.
  latOffset = 0;
  lonOffset = 0;
  // Granularity of dates, normally represented in units of milliseconds since the 1970 epoch.
  dateGranularity = 1000;

  /**
   * @param pbf - the Protobuf object to read from
   * @param reader - the OSMReader to modify
   */
  constructor(
    public pbf: Protobuf,
    public reader: OSMReader,
  ) {
    pbf.readFields(this.#readLayer, this, 0);
  }

  /**
   * Get a string from the string table at the given index
   * @param index - the index of the string in the string table
   * @returns - the string
   */
  getString(index: number): string {
    return this.stringtable.get(index);
  }

  /**
   * Get a record of strings from the string table
   * @param keys - list of indices for the keys
   * @param values - list of indices for the values
   * @returns - the record or object containing the key-value pairs
   */
  tags(keys: number[], values: number[]): Record<string, string> {
    const res: Record<string, string> = {};
    for (let i = 0; i < keys.length; i++) {
      res[this.getString(keys[i])] = this.getString(values[i]);
    }
    return res;
  }

  /**
   * Read the primitive block's contents into an object
   * @param tag - the tag of the message
   * @param pb - the primitive block to modify
   * @param pbf - the Protobuf object to read from
   */
  #readLayer(tag: number, pb: PrimitiveBlock, pbf: Protobuf): void {
    if (tag === 1) pb.stringtable = new StringTable(pbf);
    else if (tag === 2) pb.primitiveGroups.push(new PrimitiveGroup(pb, pbf));
    else if (tag === 17) pb.granularity = pbf.readVarint();
    else if (tag === 18) pb.dateGranularity = pbf.readVarint();
    else if (tag === 19) pb.latOffset = pbf.readVarint();
    else if (tag === 20) pb.lonOffset = pbf.readVarint();
    else throw new Error(`unknown tag ${tag}`);
  }
}

/** Group of OSMPrimitives. All primitives in a group must be the same type. */
export class PrimitiveGroup {
  // changesets: ChangeSet[] = [];
  /**
   * @param primitiveBlock - the parent PrimitiveBlock
   * @param pbf - the Protobuf object to read from
   */
  constructor(
    public primitiveBlock: PrimitiveBlock,
    public pbf: Protobuf,
  ) {
    pbf.readMessage(this.#readLayer, this);
  }

  /**
   * @param tag - the tag of the message
   * @param pg - the primitive group to modify
   * @param pbf - the Protobuf object to read from
   */
  #readLayer(tag: number, pg: PrimitiveGroup, pbf: Protobuf): void {
    const { primitiveBlock } = pg;
    const { reader } = primitiveBlock;
    const { skipWays, skipRelations } = reader;
    const skipWR = skipWays && skipRelations;

    if (tag === 1) {
      const node = new Node(primitiveBlock, reader, pbf);
      if (!node.isFilterable()) reader.nodes.set(node.id, node.toVectorFeature());
      if (!skipWR) reader.nodeGeometry.set(node.id, node.toVectorGeometry());
    } else if (tag === 2) {
      const dn = new DenseNodes(primitiveBlock, reader, pbf);
      for (const node of dn.nodes()) {
        if (!node.isFilterable()) reader.nodes.set(node.id, node.toVectorFeature());
        if (!skipWR) reader.nodeGeometry.set(node.id, node.toVectorGeometry());
      }
    } else if (tag === 3) {
      if (skipWR) return;
      const way = new Way(primitiveBlock, reader, pbf);
      const refs = way.nodeRefs();
      const ivf = way.toVectorFeature();
      if (refs.length >= 2) reader.wayGeometry.set(way.id, refs);
      if (ivf !== undefined && !way.isFilterable()) reader.ways.set(way.id, ivf);
    } else if (tag === 4) {
      if (skipRelations) return;
      const relation = new Relation(primitiveBlock, reader, pbf);
      const ivf = relation.toIntermediateFeature();
      if (ivf === undefined) return;
      for (const member of getNodeRelationPairs(ivf.members)) {
        reader.nodeRelationPairs.set(member.node, member);
      }
      if (!relation.isFilterable()) reader.relations.set(relation.id, ivf);
    }
    // else if (tag === 5) {
    //   this.changesets.push(new ChangeSet(pbf));
    // }
  }
}

/**
 * String table, contains the common strings in each block.
 * Note that we reserve index '0' as a delimiter, so the entry at that
 * index in the table is ALWAYS blank and unused.
 * NOTE: OSM isn't safe and allows " inside of strings, so we have to replace them with '
 * NOTE: OSM isn't safe and allows \ at the end of strings, so we have to remove them so it can be properly parsed.
 */
export class StringTable {
  strings: string[] = [];

  /**
   * @param pbf - the Protobuf object to read from
   */
  constructor(pbf: Protobuf) {
    pbf.readMessage(this.#readLayer, this);
  }

  /**
   * @param index - the index of the string
   * @returns - the string
   */
  get(index: number): string {
    return this.strings[index];
  }

  /**
   * @param tag - the tag of the message
   * @param st - the string table to modify
   * @param pbf - the Protobuf object to read from
   */
  #readLayer(tag: number, st: StringTable, pbf: Protobuf): void {
    if (tag === 1) st.strings.push(pbf.readString());
    else throw new Error(`unknown tag ${tag}`);
  }
}

/** This is kept for backwards compatibility but not used anywhere. */
// export class ChangeSet {
//   id = 0;

//   /**
//    * @param pbf
//    */
//   constructor(pbf: Protobuf) {
//     pbf.readMessage(this.#readLayer, this);
//   }

//   /**
//    * @param tag
//    * @param cs
//    * @param pbf
//    */
//   #readLayer(tag: number, cs: ChangeSet, pbf: Protobuf): void {
//     if (tag === 1) cs.id = pbf.readVarint();
//     else throw new Error(`unknown tag ${tag}`);
//   }
// }

File: rust/readers/osm/primitive.rs
use crate::data_structures::HasLayer;

use super::{
    info::InfoBlock,
    node::{DenseNodes, Node},
    relation::{IntermediateNodeMember, MemberType, Relation},
    way::Way,
};
use alloc::{string::String, vec::Vec};
use pbf::{ProtoRead, Protobuf};
use s2json::Properties;

/// A metadata struct for relations
#[derive(Debug, Clone, PartialEq)]
pub struct OSMMetadataRelation {
    /// The role of the relation related to the VectorFeature
    pub role: String,
    /// The properties of the relation
    pub properties: Properties,
}

/// The expected metadata in the VectorFeature for all types (node, way, relation)
#[derive(Debug, Clone, PartialEq)]
pub struct OSMMetadata {
    /// The type of the VectorFeature
    pub osm_type: MemberType,
    /// the info block describing the data storage history
    pub info: Option<InfoBlock>,
    /// the nodes that make up the VectorFeature
    pub nodes: Option<Vec<IntermediateNodeMember>>,
    /// if this feature is part of a relation, this relation will describe the other components
    pub relation: Option<OSMMetadataRelation>,
}
impl HasLayer for OSMMetadata {
    fn get_layer(&self) -> Option<String> {
        None
    }
}

/// NOTE: currently relations are stored, but we don't wait for the Block to store all relations
/// before we start testing primtiveHandle against the data. This is a problem because
/// relations reference eachother at times, and we need to be able to resolve those references
/// before we can run relationHandle against the data. This isn't an important issue since
/// in practice, all relations that reference eachother often produce garbage or unusable data.
/// But it would be *nice* to fix this. Morbidly enough, the "BEST" solution is to treat relations
/// like we do nodes and ways since relations could possibly reference eachother outside their own block.
/// From a practical standpoint, I can't see this being worth the effort or memory/time cost.
#[derive(Debug)]
pub struct PrimitiveBlock {
    stringtable: StringTable,
    /// Primitive groups are smaller collections of the nodes, ways and relations
    pub primitive_groups: Vec<PrimitiveGroup>,
    /// Granularity, units of nanodegrees, used to store coordinates in this block.
    pub granularity: i32,
    /// Offset value between the output coordinates and the granularity grid in units of nanodegrees.
    pub lat_offset: i64,
    /// Offset value between the output coordinates and the granularity grid in units of nanodegrees.
    pub lon_offset: i64,
    /// Granularity of dates, normally represented in units of milliseconds since the 1970 epoch.
    pub date_granularity: i32,
}
impl Default for PrimitiveBlock {
    fn default() -> Self {
        Self {
            stringtable: StringTable::default(),
            primitive_groups: Vec::new(),
            granularity: 100,
            lat_offset: 0,
            lon_offset: 0,
            date_granularity: 1000,
        }
    }
}
impl PrimitiveBlock {
    /// Get a string from the string table
    pub fn get_string(&self, index: usize) -> &str {
        self.stringtable.get(index)
    }

    /// Get a record of strings from the string table
    pub fn tags(&self, keys: &[u32], values: &[u32]) -> Properties {
        let mut res = Properties::default();
        // zip the keys and values and put them in the map
        for (key, value) in keys.iter().zip(values.iter()) {
            let key = self.get_string(*key as usize);
            let value = self.get_string(*value as usize);
            res.insert(key.into(), value.into());
        }
        res
    }
}
/// Read in the contents of the primitive block
impl ProtoRead for PrimitiveBlock {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            1 => pb.read_message(&mut self.stringtable),
            2 => {
                let mut group = PrimitiveGroup::default();
                pb.read_message(&mut group);
                self.primitive_groups.push(group);
            }
            17 => self.granularity = pb.read_varint(),
            18 => self.date_granularity = pb.read_varint(),
            19 => self.lat_offset = pb.read_varint(),
            20 => self.lon_offset = pb.read_varint(),
            _ => panic!("unknown tag {}", tag),
        }
    }
}

/// Group of OSMPrimitives. All primitives in a group must be the same type.
#[derive(Debug, Default)]
pub struct PrimitiveGroup {
    /// Nodes (points)
    pub nodes: Vec<Node>,
    /// Ways (lines and polygons)
    pub ways: Vec<Way>,
    /// Relations - collections of nodes, ways and relations
    pub relations: Vec<Relation>,
    // pub changesets: Vec<ChangeSet>,
}
/// Read in the contents of the primitive block
impl ProtoRead for PrimitiveGroup {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            1 => {
                let mut node = Node::default();
                pb.read_message(&mut node);
                self.nodes.push(node);
            }
            2 => {
                let mut dense_nodes = DenseNodes::default();
                pb.read_message(&mut dense_nodes);
                self.nodes.extend(dense_nodes.nodes());
            }
            3 => {
                let mut way = Way::default();
                pb.read_message(&mut way);
                self.ways.push(way);
            }
            4 => {
                let mut relation = Relation::default();
                pb.read_message(&mut relation);
                self.relations.push(relation);
            }
            // 5 => {
            //     let mut changeset = ChangeSet::default();
            //     pb.read_message(&mut changeset);
            //     self.changesets.push(changeset);
            // }
            _ => panic!("unknown tag {}", tag),
        }
    }
}

/// String table, contains the common strings in each block.
/// Note that we reserve index '0' as a delimiter, so the entry at that
/// index in the table is ALWAYS blank and unused.
/// NOTE: OSM isn't safe and allows " inside of strings, so we have to replace them with '
/// NOTE: OSM isn't safe and allows \ at the end of strings, so we have to remove them so it can be properly parsed.
#[derive(Debug, Default)]
pub struct StringTable {
    strings: Vec<String>,
    empty_string: String,
}
impl StringTable {
    /// Get a string from the string table
    pub fn get(&self, index: usize) -> &str {
        self.strings.get(index).unwrap_or(&self.empty_string)
    }
}
/// Read in the contents of the header block
impl ProtoRead for StringTable {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            1 => self.strings.push(pb.read_string()),
            _ => panic!("unknown tag {}", tag),
        }
    }
}

// /// This is kept for backwards compatibility but not used anywhere.
// #[derive(Debug, Default)]
// pub struct ChangeSet {
//     /// The id of the changeset
//     pub id: i64,
// }
// /// Read in the contents of the header block
// impl ProtoRead for ChangeSet {
//     fn read(&mut self, tag: u64, pb: &mut Protobuf) {
//         match tag {
//             1 => self.id = pb.read_varint(),
//             _ => panic!("unknown tag {}", tag),
//         }
//     }
// }
